### DISCUSSION TITLE
ui.log start scroll at bottom of div

### DISCUSSION THREAD

**Discussion by mattie47** [zauberzeug/nicegui#3909] [Q&A]
### Question

Hi,

I'm using the ui logger for displaying device logs. 

![image](https://github.com/user-attachments/assets/fbd35d72-c9b2-4126-86f6-f884eac05732)

As per the screenshot, this can be quite a lot of data. 

We added a "autoscroll" button a year ago which essentially just does a `ui.log.push(buffer)`.

This works well, and we're only pushing new content (as opposed to sending the entire log constantly).

**However our issue is that the scroll bar never stays at the bottom after the push of new logs** 

This means when we're watching logs in real-time, we can't be looking at the ones that have just arrived. 

Apologies that I'm not providing a huge amount of code, but this is how I'm initializing ui.log:

```py
        logObjects.canHandler.guiComponent = ui.log(max_lines=200).style('width: 100%; min-height: 266px;').classes("hover:resize-y").props("name=canLog")
 
        ui.timer(interval=0.5, callback=lambda: update_can_log(logObjects))
```

I did try a couple pieces of Javascript such as below (modified from stackoverflow [here](https://stackoverflow.com/questions/68026897/set-scroll-position-inside-div-element-to-bottom)):

```py
def update_can_log(logObjects: LogObjects):
    if logObjects.autoscroll:
        logObjects.canHandler.push_to_ui()
        ui.run_javascript(f'''
                var logDiv = document.querySelector('div[name="canLog"]');
                if (logDiv) {{logDiv.scrollTop = logDiv.scrollHeight;}}
            ''')

def refresh_can_log(canHandler: GUILoggerHandler):
    if canHandler.logs == "":
        return
    
    latest_logs = canHandler.logs.rsplit('\n', 201)
    if len(latest_logs) > 200:
        latest_logs = latest_logs[1:]
    canHandler.guiComponent.clear()
    canHandler.guiComponent.push("\n".join(latest_logs))
```

However this was quite "jumpy/jittery". and not a smooth user experience.

There was also https://stackoverflow.com/questions/18614301/keep-overflow-div-scrolled-to-bottom-unless-user-scrolls-up
which kinda worked, however I couldn't get it working with a parent div (to ensure the reversed content would reverse back to the correct order).

I'm just curious if anyone has an idea how I can be getting the logging updating from the bottom of the log window (i.e. showing the latest logs first, as they "arrive") ? 

Thanks,
Matt

**Comment by mattie47** [comment]
Hmm. I might be on to something with the following:

```py
        with ui.row().style("overflow-anchor: none; width: 100%; min-height: 266px").props("name=scroller"):
            logObjects.canHandler.guiComponent = ui.log(max_lines=200).style('overflow-anchor: none;').classes("hover:resize-y").props("name=canLog")
            ui.row().style("overflow-anchor: auto; height: 1px;").props("name=anchor")
```

which is based on https://css-tricks.com/books/greatest-css-tricks/pin-scrolling-to-bottom/

**Comment by falkoschindler** [comment]
Hi @mattie47,

I don't quite understand the difference to a simple example like this:

```py
log = ui.log(max_lines=100).classes('h-20')
ui.timer(1.0, lambda: log.push(datetime.now().strftime('%X.%f')[:-5]))
```

You can either scroll up and read the log while new data is pushed to the bottom, or scroll to the bottom and see new data arriving.

Maybe you can provide a minimum reproducible example so that we can better understand the problem.
[Why is it important to provide a minimal reproducible example?](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-it-important-to-provide-a-minimal-reproducible-example)

**Reply by golgitab** [reply]
Hi @falkoschindler,

I have just encountered the same issue as the author of this thread. 
Issue is I want to see new log every time it arrives unless I scroll from bottom to any position.

Is it possible to scroll down when scrollTop == scrollHeight and do not when it's not a case?

**Reply by golgitab** [reply]
Just wrote some code to implement what i wanted using custom ui.log:

```py
class CustomLog(ui.log):

    def __init__(self, *args):
        super().__init__(*args)
        
        self.scroll = True
        # self.id = 'custom-log'

    def push(self, line: str) -> None:
        """Add a new line to the log.

        :param line: the line to add (can contain html tags)
        """

        asyncio.create_task(self.push_async(line))


    async def scrollCheck(self):
        with self:
            scroll = await ui.run_javascript(f"""
                const logDiv = document.getElementById('{self.html_id}');
                return ((logDiv.scrollTop + logDiv.clientHeight)== logDiv.scrollHeight);
            """)
        
        if scroll:
            self.scroll = True
        else:
            self.scroll = False
    
    async def push_async(self, line):
        with self:
            await self.scrollCheck()
            
            ui.html(line)            
            
            if self.scroll:
                ui.run_javascript(f"""
                    const logDiv = document.getElementById('{self.html_id}');
                    if (logDiv) {{
                        logDiv.scrollTop = logDiv.scrollHeight;
                    }} else {{
                        console.log('nicegui-log not found');
                    }}
                """)
        while self.max_lines is not None and len(self.default_slot.children) > self.max_lines:
            self.remove(0)
```

**Reply by falkoschindler** [reply]
@golgitab I don't understand how your `CustomLog` behaves differently than `ui.log`. Maybe the current CSS solution doesn't work on your browser?
https://github.com/zauberzeug/nicegui/blob/26765aaa7ec49c6e92b3e628fbd3c5a191fa1462/nicegui/static/nicegui.css#L228-L240

**Reply by pandabearcodes** [reply]
Hi @falkoschindler 

I am currently experiencing something similar. I believe it might be related to the number of log lines being pushed within a certain time frame.

For example:

```py
from nicegui import ui
import time

log = ui.log().classes('w-full h-30vh')

def write_log():
    log.push(f'Log entry at {time.strftime("%H:%M:%S")}')
    log.push(f'Log entry at {time.strftime("%H:%M:%S")}')
    log.push(f'Log entry at {time.strftime("%H:%M:%S")}')
    log.push(f'Log entry at {time.strftime("%H:%M:%S")}')

ui.timer(0.5, write_log, active=True)

ui.run()
```

Here the autoscroll does not seem to work as expected, whilst when we reduce the log messages to:

```py
def write_log():
    log.push(f'Log entry at {time.strftime("%H:%M:%S")}')
```

it will work as expected.

This is on Chrome Version 136.0.7103.114

Thank you very much for all your efforts. nicegui is incredibly useful to me.

**Reply by falkoschindler** [reply]
Oh, thanks for bringing this up, @pandabearcodes! I haven't considered pushing multiple lines at once.
Since this can be considered a bug and I don't immediately see how to fix or workaround it, I opened a new issue #4830.
Any help is very welcome.

**Reply by golgitab** [reply]
@falkoschindler Hi,

Just to clarify my issue (in addition to what you have discovered with @pandabearcodes message):
When the scroll of the log is snapped to the end - i want auto scroll down when new log\logs is\are added, but when user, for example, scrolled in the middle of logs to read something, i expect to prevent auto scrolling down if any new logs added. That what i was trying to explain, i hope i did it better now

**Reply by pandabearcodes** [reply]
@golgitab 

just replying quickly cause i got the notification.
what you are describing is exactly what ui.log is doing currently. if you try out my example with only one log message in the function - you will see this effect. i believe that's the cause of the confusion - since you are both saying the same thing.

the issue seems to be when "too many" (however many that is) log messages are pushed at once. then this behavior stops.

**Reply by golgitab** [reply]
@pandabearcodes thanks for your reply!

Indeed it works as I described, it took me sometime to repeat my problem, because your example worked even within my code with simple text pushes, but once I tried to push pandas DataFrame, it stopped responding as needed. I guess in the end issue is the same as with multiple pushes at one time, just got it differentely
